components/common/src/feature_flags.rs

```diff
@@ -0,0 +1,97 @@
+//! the `ENV_VARS` map below.
+
+use crate::ui::{UIWriter,
+                UI};
+use features::features;
+use habitat_core::env as henv;
+use lazy_static::lazy_static;
+use std::{collections::HashMap,
+          iter::FromIterator};
+
+// TODO (CM): It would be nice to come up with a way to more
+// programmatically manage these flags. It's a bit of a pain to define
+// the flag, and then define the environment variables
+// separately. Nothing statically guarantees that you've specified an
+// variable for a flag.
+
+// TODO (CM): It'd be great to have a built-in way to document them,
+// too.
+
+// TODO (CM): Part of that documentation might be *when* a flag was
+// added. In general, long-lived flags are a code-smell.
+
+// TODO (CM): It may also be useful to break out features by area of
+// concern. The `features` crate allows us to define multiple sets of
+// feature flags and deal with them separately.
+
+features! {
+    pub mod feat {
+        const List           = 0b0000_0000_0001,
+        const TestExit       = 0b0000_0000_0010,
+        const TestBootFail   = 0b0000_0000_0100,
+        const RedactHTTP     = 0b0000_0000_1000,
+        const IgnoreSignals  = 0b0000_0001_0000,
+        const InstallHook    = 0b0000_0010_0000,
+        const OfflineInstall = 0b0000_0100_0000,
+        const IgnoreLocal    = 0b0000_1000_0000,
+        const EventStream    = 0b0001_0000_0000
+    }
+}
+
+lazy_static! {
+    static ref ENV_VARS: HashMap<feat::Flags, &'static str> = {
+        let mapping = vec![(feat::List, "HAB_FEAT_LIST"),
+                           (feat::TestExit, "HAB_FEAT_TEST_EXIT"),
+                           (feat::TestBootFail, "HAB_FEAT_BOOT_FAIL"),
+                           (feat::RedactHTTP, "HAB_FEAT_REDACT_HTTP"),
+                           (feat::IgnoreSignals, "HAB_FEAT_IGNORE_SIGNALS"),
+                           (feat::InstallHook, "HAB_FEAT_INSTALL_HOOK"),
+                           (feat::OfflineInstall, "HAB_FEAT_OFFLINE_INSTALL"),
+                           (feat::IgnoreLocal, "HAB_FEAT_IGNORE_LOCAL"),
+                           (feat::EventStream, "HAB_FEAT_EVENT_STREAM")];
+        HashMap::from_iter(mapping)
+    };
+}
+
+/// If the environment variable for a flag is set to _anything_ but
+/// the empty string, it is activated.
+pub fn enable_features_from_env(ui: &mut UI) {
+    for (feature, env_var) in ENV_VARS.iter() {
+        if henv::var(env_var).is_ok() {
+            feat::enable(*feature);
+            ui.warn(&format!("Enabling feature: {:?}", feature))
+              .unwrap();
+        }
+    }
+
+    // TODO (CM): Once the other TODOs above are done (especially the
+    // documentation bits), it would be nice to extract this logic
+    // into an actual discoverable CLI subcommand; it's a little weird
+    // that you have to know how to enable a feature flag before you
+    // can even find out that there *are* feature flags to enable.
```

Comment:
This is really a question about what the intended use of feature flags is. I've been treating them as guards around pre-release code that isn't documented and may be totally broken. If that's what they are, I think they're non-discoverable *by design*. Once a feature is documented and of standard quality, I don't see why it wouldn't just be a regular CLI argument (with optional `env`-based default) and therefore totally discoverable.

Reply:
That's kind of what I'm getting at... `HAB_FEAT_LIST` isn't "broken pre-release code"; it's clearly intended as a way to show what features are currently in-flight. It seems like a useful bit of functionality, but it _itself_ isn't a "feature" like other things are.

The few "test" features that you added are also useful, but also aren't quite "features" as we're discussing here, either.

(Also, just as an aside, just because something is currently behind a feature flag doesn't necessarily mean that once it's fully baked it will become anything that's exposed on the CLI; it might just be a new way of doing something internally)

Reply:
After thinking about it more, I think the test features I've added *shouldn't* be part of `HAB_FEAT_LIST`, they should be proper CLI options, but [hidden from the clap help](https://docs.rs/clap/2.32.0/clap/struct.Arg.html#method.hidden), because they're not user-facing.

Maybe it's acceptable to have feature flag code be potentially broken and untested, but once we have confidence in the long-term utility, they should be moved into the proper CLI (hidden or not) and integrated into our interfaces in a properly testable way.

Reply:
I guess I don't think `HAB_FEAT_LIST` should really be a thing at all. It's a little weird that that output could show up, say, as you start up `hab sup run`. It feels like the only time you'd want to enable that feature is if you were doing something like `HAB_FEAT_LIST=1 hab --help`... which implies something like `hab features` would be the clearer thing. That seems more in line with the intention of the existing code.

I think there are probably other wrinkles to this, centering around the fact that `hab` and `hab-sup` are two separate binaries, and I'm fine with punting this for later (which is why I just mentioned this in a comment).

Reply:
Having the test features be hidden CLI options is nice, though. Since the `features` crate also lets you create multiple sets of flags, we could also move them out into a separate set, apart from the "real" feature flags.

I think either way is fine.